README.md
GOSPL
=====

The GOSPL is a public collection of POPLOG source code that is reviewed and maintained by Stephen Leach.

gospl/NEWS.txt
docs/news.txt

pending/plog/lib/poem.pl
/* poem.pl								*/

/************************************************************************/
/*									*/
/* Prolog Object-oriented Embedded Manager				*/
/*									*/
/* Externally callable predicates:				 	*/
/*									*/
/*   - poem/0 translates the 'class' structures into Prolog clauses	*/
/*									*/
/*   - new/2 establishes a validated instance of a class		*/
/*									*/
/*   - instance/1 succeeds if its argument is a valid class instance	*/
/*									*/
/*   - ':'/2 (infix operator) handles class predicate accessing		*/
/*									*/
/************************************************************************/


?- op( 255, yfx, body ).
?- op( 255, xfx, checks ).
?- op( 200, xfx, class ).
?- op( 200, fx, class ).
?- op( 150, yfx, -&- ).
?- op( 100, xfx, => ).
?- op( 60, fx, not ).
?- op( 5, xfx, : ).


/* poem/0 is the top-level callable predicate to translate the class	*/
/* patterns into Prolog predicates.					*/

poem :-
	subclass_translations,
	class_translations.
	/* The class patterns (functor 'body') could be retracted now	*/


/* Subclass_translations/0 preprocesses subclass definitions for later	*/
/* processing by class_translations.					*/

/* Subclasses access their superclass predicates by:			*/
/*	subclassname(X) :- superclassname(X).				*/
/* The new subclass predicates are bound into a new class definition	*/
/* along with a check that the subclass object is a valid instance of	*/
/* the superclass, by calling instance(superclassname(arguments)).	*/

subclass_translations :-
	( SubCDesc class SuperCName checks Conds body Body ),
	SubCDesc =.. [ SubCName|Args ],
	SuperCDesc =.. [ SuperCName|Args ],
	SubCCall =.. [ SubCName,X ],
	SuperCCall =.. [ SuperCName,X ],
	assertz( (SubCCall :- SuperCCall) ),
	assertz( (class SubCDesc
		      checks
			  (instance(SuperCDesc), Conds)
		      body Body) ),
	fail.

subclass_translations.


/* class_translations/0 splits the class patterns into Prolog		*/
/* predicates for the defined class operations.				*/

class_translations :- 
	( class Name checks Conds body Preds ),
	assert_checks( Name, Conds ),
	Name =.. [ CName|CArgs ],
	assert_predicates( CName, Preds, [CArgs] ),
	fail.

class_translations.


/* assert_checks/2 translates the initialisation list of a class into	*/
/* a Prolog clause instance/1, which checks that its argument is a 	*/
/* valid instance of its class.						*/

assert_checks( ClassName, Checks ) :-
	assertz( instance(ClassName) :- Checks ).

none.	/* so that classes defined with "checks none" work properly.	*/


/* assert_predicates/3 translates the predicate list of a class into	*/
/* correct Prolog clauses, bound with the class name for uniqueness.	*/
/* To access the owner object variables, they are bound in as a final	*/
/* argument to each predicate.						*/

assert_predicates( _, none, _ ) :-
	!.
  
assert_predicates( CName, P1 -&- P2, Args ) :-
	assert_predicates( CName, P1, Args ),
	assert_predicates( CName, P2, Args ).

assert_predicates( CName, Head => Tail, Args ) :-
	Head =.. [ HFunc|HArgs ],
	append( HArgs, Args, NewArgs ),		/* add object descrn.	*/
	NewHead =.. [ HFunc|NewArgs ],
	UniqueHead =.. [ CName,NewHead ],	/* tag with class name	*/
	assertz( (UniqueHead :- Tail) ).


/* new/2 instantiates a new object and performs any constraint checks	*/
/*   eg new( point(3,4), Point34 ).					*/

new( Classinstance, Classinstance ) :-
	instance( Classinstance ).


/* ':' is an operator to select predicate application according to	*/
/* object class.  For example,						*/
/*   if			PointA = point(1, 2),				*/
/*   then		PointA:distance(PointB, D)			*/
/*   translates to	point( distance(PointB, D, 1, 2) )		*/

ClassInst : Function :-
	ClassInst =.. [ CFunc|CArgs ],
	Function =.. [ FFunc|FArgs ],
	append( FArgs, [CArgs], NewArgs ),
	NewHead =.. [ FFunc|NewArgs ],
	UniqueHead =.. [ CFunc,NewHead ],
	!,		/* no point resatisfying the translation	*/
	call( UniqueHead ).


/* Finally, standard append/3 definition .				*/

append([], X, X).

append([X|L1], L2, [X|L3]) :-
	append(L1, L2, L3).


pending/dbm/gdbm/falloc.c
/* falloc.c - The file space management routines for dbm. */

/*  This file is part of GDBM, the GNU data base manager, by Philip A. Nelson.
    Copyright (C) 1990, 1991  Free Software Foundation, Inc.

    GDBM is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 1, or (at your option)
    any later version.

    GDBM is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with GDBM; see the file COPYING.  If not, write to
    the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

    You may contact the author by:
       e-mail:  phil@cs.wwu.edu
      us-mail:  Philip A. Nelson
                Computer Science Department
                Western Washington University
                Bellingham, WA 98226
        phone:  (206) 676-3035
       
*************************************************************************/

#include "gdbmdefs.h"


/* The forward definitions for this file.  See the functions for
   the definition of the function. */

static avail_elem get_elem ();
static avail_elem get_block ();
static push_avail_block ();
static pop_avail_block ();
static adjust_bucket_avail ();

/* Allocate space in the file DBF for a block NUM_BYTES in length.  Return
   the file address of the start of the block.  

   Each hash bucket has a fixed size avail table.  We first check this
   avail table to satisfy the request for space.  In most cases we can
   and this causes changes to be only in the current hash bucket.
   Allocation is done on a first fit basis from the entries.  If a
   request can not be satisfied from the current hash bucket, then it is
   satisfied from the file header avail block.  If nothing is there that
   has enough space, another block at the end of the file is allocated
   and the unused portion is returned to the avail block.  This routine
   "guarantees" that an allocation does not cross a block boundary unless
   the size is larger than a single block.  The avail structure is
   changed by this routine if a change is needed.  If an error occurs,
   the value of 0 will be returned.  */

long
_gdbm_alloc (dbf, num_bytes)
     gdbm_file_info *dbf;
     int num_bytes;
{
  long  file_adr;		/* The address of the block. */
  avail_elem av_el;		/* For temporary use. */

  /* The current bucket is the first place to look for space. */
  av_el = get_elem (num_bytes, dbf->bucket->bucket_avail,
		    &dbf->bucket->av_count);

  /* If we did not find some space, we have more work to do. */
  if (av_el.av_size == 0)
    {
     /* Is the header avail block empty and there is something on the stack. */
      if (dbf->header->avail.count == 0 && dbf->header->avail.next_block != 0)
	pop_avail_block (dbf);
      
      /* Get another full block from end of file. */
      av_el = get_block (num_bytes, dbf);

      dbf->header_changed = TRUE;
    }

  /* We now have the place from which we will allocate the new space. */
  file_adr = av_el.av_adr;

  /* Put the unused space back in the avail block. */
  av_el.av_adr += num_bytes;
  av_el.av_size -= num_bytes;
  _gdbm_free (dbf, av_el.av_adr, av_el.av_size);

  /* Return the address. */
  return file_adr;
  
}



/* Free space of size NUM_BYTES in the file DBF at file address FILE_ADR.  Make
   it avaliable for reuse through _gdbm_alloc.  This routine changes the
   avail structure.  The value TRUE is returned if there were errors.  If no
   errors occured, the value FALSE is returned. */

_gdbm_free (dbf, file_adr, num_bytes)
     gdbm_file_info *dbf;
     long file_adr;
     int num_bytes;
{
  avail_elem temp;

  /* Is it too small to worry about? */
  if (num_bytes <= IGNORE_SIZE)
    return;

  /* Initialize the avail element. */
  temp.av_size = num_bytes;
  temp.av_adr = file_adr;

  /* Is the freed space large or small? */
  if (num_bytes >= dbf->header->block_size)
    {
      if (dbf->header->avail.count == dbf->header->avail.size)
	{
	  push_avail_block (dbf);
	}
      _gdbm_put_av_elem (temp, dbf->header->avail.av_table,
			 &dbf->header->avail.count);
      dbf->header_changed = TRUE;
    }
  else
    {
      /* Try to put into the current bucket. */
      if (dbf->bucket->av_count < BUCKET_AVAIL)
	_gdbm_put_av_elem (temp, dbf->bucket->bucket_avail,
			   &dbf->bucket->av_count);
      else
	{
	  if (dbf->header->avail.count == dbf->header->avail.size)
	    {
	      push_avail_block (dbf);
	    }
	  _gdbm_put_av_elem (temp, dbf->header->avail.av_table,
			     &dbf->header->avail.count);
	  dbf->header_changed = TRUE;
	}
    }

  if (dbf->header_changed)
    adjust_bucket_avail (dbf);

  /* All work is done. */
  return;
}



/* The following are all utility routines needed by the previous two. */


/* Gets the avail block at the top of the stack and loads it into the
   active avail block.  It does a "free" for itself! */

static
pop_avail_block (dbf)
     gdbm_file_info *dbf;
{
  int  num_bytes;		/* For use with the read system call. */
  long file_pos;		/* For use with the lseek system call. */
  avail_elem temp;

  /* Set up variables. */
  temp.av_adr = dbf->header->avail.next_block;
  temp.av_size = ( ( (dbf->header->avail.size * sizeof (avail_elem)) >> 1)
		  + sizeof (avail_block));

  /* Read the block. */
  file_pos = lseek (dbf->desc, temp.av_adr, L_SET);
  if (file_pos != temp.av_adr)  _gdbm_fatal (dbf, "lseek error");
  num_bytes = read (dbf->desc, &dbf->header->avail, temp.av_size);
  if (num_bytes != temp.av_size) _gdbm_fatal (dbf, "read error");

  /* We changed the header. */
  dbf->header_changed = TRUE;

  /* Free the previous avail block. */
  _gdbm_put_av_elem (temp, dbf->header->avail.av_table,
		     &dbf->header->avail.count);
}


/* Splits the header avail block and pushes half onto the avail stack. */

static
push_avail_block (dbf)
     gdbm_file_info *dbf;
{
  int  num_bytes;
  int  av_size;
  int  av_adr;
  int  index;
  long file_pos;
  avail_block *temp;
  avail_elem  new_loc;
 

  /* Caclulate the size of the split block. */
  av_size = ( (dbf->header->avail.size * sizeof (avail_elem)) >> 1)
            + sizeof (avail_block);

  /* Get address in file for new av_size bytes. */
  new_loc = get_elem (av_size, dbf->header->avail.av_table,
		      &dbf->header->avail.count);
  if (new_loc.av_size == 0)
    new_loc = get_block (av_size, dbf);
  av_adr = new_loc.av_adr;


  /* Split the header block. */
  temp = (avail_block *) alloca (av_size);
  /* Set the size to be correct AFTER the pop_avail_block. */
  temp->size = dbf->header->avail.size;
  temp->count = 0;
  temp->next_block = dbf->header->avail.next_block;
  dbf->header->avail.next_block = av_adr;
  for (index = 1; index < dbf->header->avail.count; index++)
    if ( (index & 0x1) == 1)	/* Index is odd. */
      temp->av_table[temp->count++] = dbf->header->avail.av_table[index];
    else
      dbf->header->avail.av_table[index>>1]
	= dbf->header->avail.av_table[index];

  /* Update the header avail count to previous size divided by 2. */
  dbf->header->avail.count >>= 1;

  /* Free the unneeded space. */
  new_loc.av_adr += av_size;
  new_loc.av_size -= av_size;
  _gdbm_free (dbf, new_loc.av_adr, new_loc.av_size);

  /* Update the disk. */
  file_pos = lseek (dbf->desc, av_adr, L_SET);
  if (file_pos != av_adr) {
ALLOCA_FREE(temp);
	_gdbm_fatal (dbf, "lseek error");
  }
  num_bytes = write (dbf->desc, temp, av_size);
  if (num_bytes != av_size) {
ALLOCA_FREE(temp);
	_gdbm_fatal (dbf, "write error");
  }

ALLOCA_FREE(temp);
}



/* Get_elem returns an element in the AV_TABLE block which is
   larger than SIZE.  AV_COUNT is the number of elements in the
   AV_TABLE.  If an item is found, it extracts it from the AV_TABLE
   and moves the other elements up to fill the space. If no block is 
   found larger than SIZE, get_elem returns a size of zero.  This
   routine does no I/O. */

static avail_elem
get_elem (size, av_table, av_count)
     int size;
     avail_elem av_table[];
     int *av_count;
{
  int index;			/* For searching through the avail block. */
  avail_elem val;		/* The default return value. */

  /* Initialize default return value. */
  val.av_adr = 0;
  val.av_size = 0;

  /* Search for element.  List is sorted by size. */
  index = 0;
  while (index < *av_count && av_table[index].av_size < size)
    {
      index++;
    }

  /* Did we find one of the right size? */
  if (index >= *av_count)
    return val;

  /* Ok, save that element and move all others up one. */
  val = av_table[index];
  *av_count -= 1;
  while (index < *av_count)
    {
      av_table[index] = av_table[index+1];
      index++;
    }

  return val;
}


/* This routine inserts a single NEW_EL into the AV_TABLE block in
   sorted order. This routine does no I/O. */

_gdbm_put_av_elem (new_el, av_table, av_count)
     avail_elem new_el;
     avail_elem av_table[];
     int *av_count;
{
  int index;			/* For searching through the avail block. */
  int index1;

  /* Is it too small to deal with? */
  if (new_el.av_size <= IGNORE_SIZE)
    return FALSE; 

  /* Search for place to put element.  List is sorted by size. */
  index = 0;
  while (index < *av_count && av_table[index].av_size < new_el.av_size)
    {
      index++;
    }

  /* Move all others up one. */
  index1 = *av_count-1;
  while (index1 >= index)
    {
      av_table[index1+1] = av_table[index1];
      index1--;
    }
  
  /* Add the new element. */
  av_table[index] = new_el;

  /* Increment the number of elements. */
  *av_count += 1;  

  return TRUE;
}



/* Get_block "allocates" new file space and the end of the file.  This is
   done in integral block sizes.  (This helps insure that data smaller than
   one block size is in a single block.)  Enough blocks are allocated to
   make sure the number of bytes allocated in the blocks is larger than SIZE.
   DBF contains the file header that needs updating.  This routine does
   no I/O.  */

static avail_elem
get_block (size, dbf)
     int size;
     gdbm_file_info *dbf;
{
  avail_elem val;

  /* Need at least one block. */
  val.av_adr  = dbf->header->next_block;
  val.av_size = dbf->header->block_size;

  /* Get enough blocks to fit the need. */
  while (val.av_size < size)
    val.av_size += dbf->header->block_size;

  /* Update the header and return. */
  dbf->header->next_block += val.av_size;

  /* We changed the header. */
  dbf->header_changed = TRUE;

  return val;
  
}


/*  When the header already needs writing, we can make sure the current
    bucket has its avail block as close to 1/2 full as possible. */
static
adjust_bucket_avail (dbf)
     gdbm_file_info *dbf;
{
  int third = BUCKET_AVAIL / 3;
  avail_elem av_el;

  /* Can we add more entries to the bucket? */
  if (dbf->bucket->av_count < third)
    {
      if (dbf->header->avail.count > 0)
	{
	  dbf->header->avail.count -= 1;
	  av_el = dbf->header->avail.av_table[dbf->header->avail.count];
	  _gdbm_put_av_elem (av_el, dbf->bucket->bucket_avail,
			     &dbf->bucket->av_count);
	  dbf->bucket_changed = TRUE;
	}
      return;
    }

  /* Is there too much in the bucket? */
  while (dbf->bucket->av_count > BUCKET_AVAIL-third
	 && dbf->header->avail.count < dbf->header->avail.size)
    {
      av_el = get_elem (0, dbf->bucket->bucket_avail, &dbf->bucket->av_count);
      _gdbm_put_av_elem (av_el, dbf->header->avail.av_table,
			 &dbf->header->avail.count);
      dbf->bucket_changed = TRUE;
    }
}

/*
JM, 9/6/93
	Added ALLOCA_FREE for systems that don't have alloca
*/


pending/dbm/gdbm/pop_gdbm.c
/* --- Copyright University of Sussex 1991. All rights reserved. ----------
 > File:            $poplocal/local/dbm/extern/pop_dbm.c
 > Purpose:         Interface to the UNIX NDBM library
 > Author:          Jonathan Meyer, Nov 10 1992
 > Documentation:   REF *GDBM_FILE, MAN * NDBM
 */

/*
Unfortunately, the GDBM library takes and returns datum structures, rather
than pointers to these structures. This means we must write a set of `C'
wrappers to the library.
*/

#include "gdbm.h"

int pop_gdbm_firstkey(dbm, dat)
GDBM_FILE *dbm;
datum *dat;
{
	/* use dbm_firstkey to fill in the user supplied datum */
	*dat = gdbm_firstkey(dbm);
	return(dat->dptr ? 0 : -1);
}

int pop_gdbm_nextkey(dbm, dat)
GDBM_FILE *dbm;
datum *dat;
{
	/* use dbm_nextkey to fill in the user supplied datum */
	*dat = gdbm_nextkey(dbm, *dat);
	return(dat->dptr ? 0 : -1);
}

int pop_gdbm_fetch(dbm, key, klen, contentd)
GDBM_FILE *dbm;
char *key;
int klen;
datum *contentd;
{
	datum keyd;
	/* fill in a datum with the key/length and then call dbm_fetch,
       setting the supplied contentd to the result.
    */
	keyd.dptr = key;
	keyd.dsize = klen;
	*contentd = gdbm_fetch(dbm, keyd);
	return(contentd->dptr ? 0 : -1);
}

int pop_gdbm_store(dbm, key, klen, content, clen)
GDBM_FILE *dbm;
char *key, *content;
int klen, clen;
{
	datum keyd, contentd;
	/* fill in a datum with the key/length and content/length, and
       then call dbm_store with update mode GDBM_REPLACE.
    */
	keyd.dptr = key;
	keyd.dsize = klen;
	contentd.dptr = content;
	contentd.dsize = clen;
	return gdbm_store(dbm, keyd, contentd, GDBM_REPLACE);
}

int pop_gdbm_delete(dbm, key, klen)
GDBM_FILE *dbm;
char *key;
int klen;
{
	datum keyd, contentd;
	keyd.dptr = key;
	keyd.dsize = klen;
	return gdbm_delete(dbm, keyd);
}

/* locates an integer in an intvec. Returns the index to the integer where
1 is the first element, or -1 if it is not found
*/
int pop_gdbm_locint(i, ivec, len)
register int i, *ivec;
int len;
{
	register int count = len;
	do {
		if (*ivec++ == i)
			return(len - count + 1);
	} while (count--);
	/* not found */
	return(-1);
}

/*
JM, 17/6/93
	Changed pop_dbm_nextkey to pass the datum it is given to
	gdbm_nextkey
*/


pending/dbm/gdbm/gdbmopen.c
/* gdbmopen.c - Open the dbm file and initialize data structures for use. */

/*  This file is part of GDBM, the GNU data base manager, by Philip A. Nelson.
    Copyright (C) 1990, 1991  Free Software Foundation, Inc.

    GDBM is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 1, or (at your option)
    any later version.

    GDBM is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with GDBM; see the file COPYING.  If not, write to
    the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

    You may contact the author by:
       e-mail:  phil@cs.wwu.edu
      us-mail:  Philip A. Nelson
                Computer Science Department
                Western Washington University
                Bellingham, WA 98226
        phone:  (206) 676-3035

*************************************************************************/


#include "gdbmdefs.h"
#include "gdbmerrno.h"

extern gdbm_error gdbm_errno;



/* Initialize dbm system.  FILE is a pointer to the file name.  If the file
   has a size of zero bytes, a file initialization procedure is performed,
   setting up the initial structure in the file.  BLOCK_SIZE is used during
   initialization to determine the size of various constructs.  If the value
   is less than 512, the file system blocksize is used, otherwise the value
   of BLOCK_SIZE is used.  BLOCK_SIZE is ignored if the file has previously
   initialized.  If READ_WRITE is set to GDBM_READ the user wants to just
   read the database and any call to dbm_store or dbm_delete will fail. Many
   readers can access the database at the same time.  If READ_WRITE is set to
   GDBM_WRITE, the user wants both read and write access to the database and
   requires exclusive access.  If READ_WRITE is GDBM_WRCREAT, the user wants
   both read and write access to the database and if the database does not
   exist, create a new one.  If READ_WRITE is GDBM_NEWDB, the user want a
   new database created, regardless of whether one existed, and wants read
   and write access to the new database.  Any error detected will cause a
   return value of null and an approprate value will be in gdbm_errno.  If
   no errors occur, a pointer to the "gdbm file descriptor" will be
   returned. */


gdbm_file_info *
gdbm_open (file, block_size, read_write, mode, fatal_func)
     char *file;
     int  block_size;
     int  read_write;
     int  mode;
     void (*fatal_func) ();
{
  gdbm_file_info *dbf;		/* The record to return. */
  struct stat file_stat;	/* Space for the stat information. */
  int         len;		/* Length of the file name. */
  int         num_bytes;	/* Used in reading and writing. */
  long        file_pos;		/* Used with seeks. */
  int	      lock_val;         /* Returned by the flock call. */
  int	      file_block_size;	/* Block size to use for a new file. */
  int 	      index;		/* Used as a loop index. */
  char        need_trunc;	/* Used with GDBM_NEWDB and locking to avoid
				   truncating a file from under a reader. */

  /* Allocate new info structure. */
  dbf = (gdbm_file_info *) malloc (sizeof (gdbm_file_info));
  if (dbf == NULL)
    {
      gdbm_errno = GDBM_MALLOC_ERROR;
      return NULL;
    }

  /* Initialize some fields for known values.  This is done so gdbm_close
     will work if called before allocating some structures. */
  dbf->dir  = NULL;
  dbf->bucket = NULL;
  for (index = 0; index < CACHE_SIZE; index++)
    {
      dbf->bucket_cache[index].ca_bucket = NULL;
      dbf->bucket_cache[index].ca_data.dptr = NULL;
    }

  /* Save name of file. */
  len = strlen (file);
  dbf->name = (char *) malloc (len + 1);
  if (dbf->name == NULL)
    {
      free (dbf);
      gdbm_errno = GDBM_MALLOC_ERROR;
      return NULL;
    }
  strcpy (dbf->name, file);

  /* Initialize the fatal error routine. */
  dbf->fatal_err = fatal_func;

  /* Open the file. */
  need_trunc = FALSE;
  if (read_write == GDBM_READER)
    {
      dbf->desc = open (dbf->name, O_RDONLY, 0);
    }
  else if (read_write == GDBM_WRITER)
    {
      dbf->desc = open (dbf->name, O_RDWR, 0);
    }
  else if (read_write == GDBM_NEWDB)
    {
      dbf->desc = open (dbf->name, O_RDWR|O_CREAT, mode);
      read_write = GDBM_WRITER;
      need_trunc = TRUE;
    }
  else
    {
      dbf->desc = open (dbf->name, O_RDWR|O_CREAT, mode);
      read_write = GDBM_WRITER;
    }
  if (dbf->desc < 0)
    {
      free (dbf->name);
      free (dbf);
      gdbm_errno = GDBM_FILE_OPEN_ERROR;
      return NULL;
    }

  /* Get the status of the file. */
  fstat (dbf->desc, &file_stat);

  /* Lock the file in the approprate way. */
  if (read_write == GDBM_READER)
    {
      if (file_stat.st_size == 0)
	{
	  close (dbf->desc);
	  free (dbf->name);
	  free (dbf);
	  gdbm_errno = GDBM_EMPTY_DATABASE;
	  return NULL;
	}
      /* Sets lock_val to 0 for success.  See systems.h. */
      READLOCK_FILE(dbf);
    }
  else
    {
      /* Sets lock_val to 0 for success.  See systems.h. */
      WRITELOCK_FILE(dbf);
    }
  if (lock_val != 0)
    {
      close (dbf->desc);
      free (dbf->name);
      free (dbf);
      if (read_write == GDBM_READER)
	gdbm_errno = GDBM_CANT_BE_READER;
      else
	gdbm_errno = GDBM_CANT_BE_WRITER;
      return NULL;
    }

  /* Record the kind of user. */
  dbf->read_write = read_write;

  /* If we do have a write lock and it was a GDBM_NEWDB, it is
     now time to truncate the file. */
  if (need_trunc && file_stat.st_size != 0)
    {
      TRUNCATE (dbf);
      fstat (dbf->desc, &file_stat);
    }

  /* Decide if this is a new file or an old file. */
  if (file_stat.st_size == 0)
    {

      /* This is a new file.  Create an empty database.  */

      /* Start with the blocksize. */
      if (block_size < 512)
	file_block_size = STATBLKSIZE;
      else
	file_block_size = block_size;

      /* Get space for the file header. */
      dbf->header = (gdbm_file_header *) malloc (file_block_size);
      if (dbf->header == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}

      /* Set the magic number and the block_size. */
      dbf->header->header_magic = 0x13579ace;
      dbf->header->block_size = file_block_size;

      /* Create the initial hash table directory.  */
      dbf->header->dir_size = 8 * sizeof (int);
      dbf->header->dir_bits = 3;
      while (dbf->header->dir_size < dbf->header->block_size)
	{
	  dbf->header->dir_size <<= 1;
	  dbf->header->dir_bits += 1;
	}

      /* Check for correct block_size. */
      if (dbf->header->dir_size != dbf->header->block_size)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_BLOCK_SIZE_ERROR;
	  return NULL;
	}

      /* Allocate the space for the directory. */
      dbf->dir = (int *) malloc (dbf->header->dir_size);
      if (dbf->dir == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}
      dbf->header->dir = dbf->header->block_size;

      /* Create the first and only hash bucket. */
      dbf->header->bucket_elems =
	(dbf->header->block_size - sizeof (hash_bucket))
	/ sizeof (bucket_element) + 1;
      dbf->header->bucket_size  = dbf->header->block_size;
      dbf->bucket = (hash_bucket *) (alloca (dbf->header->bucket_size));
      if (dbf->bucket == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}
      _gdbm_new_bucket (dbf, dbf->bucket, 0);
      dbf->bucket->av_count = 1;
      dbf->bucket->bucket_avail[0].av_adr = 3*dbf->header->block_size;
      dbf->bucket->bucket_avail[0].av_size = dbf->header->block_size;

      /* Set table entries to point to hash buckets. */
      for (index = 0; index < dbf->header->dir_size / sizeof (int); index++)
	dbf->dir[index] = 2*dbf->header->block_size;

      /* Initialize the active avail block. */
      dbf->header->avail.size
	= ( (dbf->header->block_size - sizeof (gdbm_file_header))
	 / sizeof (avail_elem)) + 1;
      dbf->header->avail.count = 0;
      dbf->header->avail.next_block = 0;
      dbf->header->next_block  = 4*dbf->header->block_size;

      /* Write initial configuration to the file. */
      /* Block 0 is the file header and active avail block. */
      num_bytes = write (dbf->desc, dbf->header, dbf->header->block_size);
      if (num_bytes != dbf->header->block_size)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_WRITE_ERROR;
ALLOCA_FREE(dbf->bucket);
	  return NULL;
	}

      /* Block 1 is the initial bucket directory. */
      num_bytes = write (dbf->desc, dbf->dir, dbf->header->dir_size);
      if (num_bytes != dbf->header->dir_size)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_WRITE_ERROR;
ALLOCA_FREE(dbf->bucket);
	  return NULL;
	}

      /* Block 2 is the only bucket. */
      num_bytes = write (dbf->desc, dbf->bucket, dbf->header->bucket_size);
      if (num_bytes != dbf->header->bucket_size)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_WRITE_ERROR;
ALLOCA_FREE(dbf->bucket);
	  return NULL;
	}

      /* Wait for initial configuration to be written to disk. */
      fsync (dbf->desc);
ALLOCA_FREE(dbf->bucket);

    }
  else
    {
      /* This is an old database.  Read in the information from the file
	 header and initialize the hash directory. */

      gdbm_file_header partial_header;  /* For the first part of it. */

      /* Read the partial file header. */
      num_bytes = read (dbf->desc, &partial_header, sizeof (gdbm_file_header));
      if (num_bytes != sizeof (gdbm_file_header))
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_READ_ERROR;
	  return NULL;
	}

      /* Is the magic number good? */
      if (partial_header.header_magic != 0x13579ace)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_BAD_MAGIC_NUMBER;
	  return NULL;
	}

      /* It is a good database, read the entire header. */
      dbf->header = (gdbm_file_header *) malloc (partial_header.block_size);
      if (dbf->header == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}
      bcopy (&partial_header, dbf->header, sizeof (gdbm_file_header));
      num_bytes = read (dbf->desc, &dbf->header->avail.av_table[1],
			dbf->header->block_size-sizeof (gdbm_file_header));
      if (num_bytes != dbf->header->block_size-sizeof (gdbm_file_header))
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_READ_ERROR;
	  return NULL;
	}
	
      /* Allocate space for the hash table directory.  */
      dbf->dir = (int *) malloc (dbf->header->dir_size);
      if (dbf->dir == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}

      /* Read the hash table directory. */
      file_pos = lseek (dbf->desc, dbf->header->dir, L_SET);
      if (file_pos != dbf->header->dir)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_SEEK_ERROR;
	  return NULL;
	}

      num_bytes = read (dbf->desc, dbf->dir, dbf->header->dir_size);
      if (num_bytes != dbf->header->dir_size)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_FILE_READ_ERROR;
	  return NULL;
	}

    }

  /* Initialize the bucket cache. */
  for (index = 0; index < CACHE_SIZE; index++)
    {
      dbf->bucket_cache[index].ca_bucket
	= (hash_bucket *) malloc (dbf->header->bucket_size);
      if (dbf->bucket_cache[index].ca_bucket == NULL)
	{
	  gdbm_close (dbf);
	  gdbm_errno = GDBM_MALLOC_ERROR;
	  return NULL;
	}
      dbf->bucket_cache[index].ca_adr = 0;
      dbf->bucket_cache[index].ca_changed = FALSE;
      dbf->bucket_cache[index].ca_data.hash_val = -1;
      dbf->bucket_cache[index].ca_data.elem_loc = -1;
    }

  /* Finish initializing dbf. */
  dbf->last_read = -1;
  dbf->bucket = dbf->bucket_cache[0].ca_bucket;
  dbf->bucket_dir = 0;
  dbf->cache_entry = &dbf->bucket_cache[0];
  dbf->header_changed = FALSE;
  dbf->directory_changed = FALSE;
  dbf->bucket_changed = FALSE;
  dbf->second_changed = FALSE;

  /* Everything is fine, return the pointer to the file
     information structure.  */
  return dbf;

}

/*
JM - 9/6/93
	Added ALLOCA_FREE for systems that don't have alloca
*/
